### PR TITLE
REST: Fix PATCH method

### PR DESCRIPTION
**Pull request by isinyaaa**
## Description
Use a converter to reconcile the PATCH body with the existing instance.

Fixes https://issues.redhat.com/browse/RHOAIENG-8534

## How Has This Been Tested?

Refer to reproducers in the Jira.

## Merge criteria:

- [x] The commits and have meaningful messages; the author will squash them [after approval](https://github.com/opendatahub-io/opendatahub-community/blob/main/contributor-cheatsheet.md#:~:text=Usually%20this%20is%20done%20in%20last%20phase%20of%20a%20PR%20revision) or will ask to merge with squash.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

cc @tarilabs

### CODE REVIEW

**Review by tarilabs — COMMENTED** [review]
/lgtm
and thank you so much @isinyaaa 

before approving one remaining question: since this was raised by UI team, would you think an update to RobotFramework with a user-story dedicated to using Patch with this PR, on in a subsequent PR, please?

**Comment by isinyaaa** [comment]
> /lgtm and thank you so much @isinyaaa
> 
> before approving one remaining question: since this was raised by UI team, would you think an update to RobotFramework with a user-story dedicated to using Patch with this PR, on in a subsequent PR, please?

that's surely a good idea, will look into it asap :)

**Comment by isinyaaa** [comment]
Added a simple test for the use-case. wdyt @tarilabs ?

File: Makefile [hunk 1]
```diff
@@ -130,7 +130,7 @@ bin/golangci-lint:
 
 GOVERTER ?= ${PROJECT_BIN}/goverter
 bin/goverter:
-	GOBIN=$(PROJECT_PATH)/bin ${GO} install github.com/jmattheis/goverter/cmd/goverter@v1.1.1
+	GOBIN=$(PROJECT_PATH)/bin ${GO} install github.com/jmattheis/goverter/cmd/goverter@v1.4.1
```
Comment (tarilabs) [inline Makefile:133]: bumps goverter 👍

Comment (tarilabs) [reply Makefile:133]: we should eventually look into avoiding specifying the version (it was required for previous work by @lampajr during some ad-interim) ... would you mind @isinyaaa logging an issue/jira to eventually change this to use a latest or some version specified in gomod if it's possible to just consolidate the version please? wdyt? (in a subsequent task of course)

File: internal/converter/openapi_converter.go [hunk 2]
```diff
@@ -50,43 +51,43 @@ type OpenAPIConverter interface {
...
 	// Ignore all fields that ARE editable
-	// goverter:default InitInferenceServiceWithUpdate
+	// goverter:default InitWithUpdate
 	// goverter:autoMap Existing
 	// goverter:ignore Id CreateTimeSinceEpoch LastUpdateTimeSinceEpoch Description ExternalId CustomProperties ModelVersionId Runtime DesiredState
 	OverrideNotEditableForInferenceService(source OpenapiUpdateWrapper[openapi.InferenceService]) (openapi.InferenceService, error)
 
 	// Ignore all fields that ARE editable
-	// goverter:default InitServeModelWithUpdate
+	// goverter:default InitWithUpdate
```
Comment (tarilabs) [inline internal/converter/openapi_converter.go:90]: make use of the generic version 👍

File: internal/converter/openapi_converter_util.go [hunk 3]
```diff
@@ -28,51 +28,14 @@ func NewOpenapiUpdateWrapper[
...
-	if source.Update != nil {
-		return *source.Update
-	}
-	return openapi.InferenceService{}
+func InitWithExisting[M OpenAPIModel](source OpenapiUpdateWrapper[M]) M {
+	return *source.Existing
 }
 
-func InitServeModelWithUpdate(source OpenapiUpdateWrapper[openapi.ServeModel]) openapi.ServeModel {
+func InitWithUpdate[M OpenAPIModel](source OpenapiUpdateWrapper[M]) M {
```
Comment (tarilabs) [inline internal/converter/openapi_converter_util.go:35]: brings this one to the generic shape too, great catch 👍

File: internal/converter/openapi_reconciler.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+// NOTE: methods must follow these patterns, otherwise tests could not find possible issues:
+// Converters patch fields entity: UpdateExisting<ENTITY>
+
+// goverter:converter
+// goverter:output:file ./generated/openapi_reconciler.gen.go
+// goverter:wrapErrors
+// goverter:enum:unknown @error
+// goverter:matchIgnoreCase
+// goverter:useZeroValueOnPointerInconsistency
+type OpenAPIReconciler interface {
```
Comment (tarilabs) [inline internal/converter/openapi_reconciler.go:14]: this "reconciler" could have also been named "updater" and the scope is given an existing (from MLMD db) entity, applies the "patch" as in the valorized fields values, returning the "to-be" state which is what the Core layer but most-importantly MLMD requires 👍

File: internal/server/openapi/api_model_registry_service_service.go [hunk 5]
```diff
@@ -450,7 +452,16 @@ func (s *ModelRegistryServiceAPIService) UpdateInferenceService(ctx context.Cont
...
+	existing, err := s.coreApi.GetInferenceServiceById(inferenceserviceId)
+	if err != nil {
+		status := api.ErrToStatus(err)
+		return Response(status, model.Error{Message: err.Error()}), nil
+	}
+	update, err := s.reconciler.UpdateExistingInferenceService(converter.NewOpenapiUpdateWrapper(existing, entity))
+	if err != nil {
+		return Response(http.StatusBadRequest, model.Error{Message: err.Error()}), nil
+	}
+	result, err := s.coreApi.UpsertInferenceService(&update)
```
Comment (tarilabs) [inline internal/server/openapi/api_model_registry_service_service.go:464]: pattern using the "reconciler" is applied to MR logical model entities:

1. get from MLMD the current entity value
2. apply the reconciler, that means applying the "patch" deltas to the existing value
3. upsert the "to-be" value back into MLMD, via the core layer upsert.

👍

File: test/robot/UserStory.robot [hunk 6]
```diff
@@ -41,6 +41,15 @@ As a MLOps engineer I would like to store a description of the model
     ${r}  Then I findModelArtifact by name and parentResourceId    name=ModelArtifactName  parentResourceId=${vId}
           And Should be equal    ${r["description"]}    sed do eiusmod tempor incididunt
 
+As a MLOps engineer I would like to update a description of the model
+    Set To Dictionary    ${registered_model}    description=Lorem ipsum dolor sit amet  name=${name}  owner=me
+    ${rId}  Given I create a RegisteredModel    payload=${registered_model}
+    &{registered_model_update}  Create dictionary    description=New description
+    ${rId}  And I update a RegisteredModel    id=${rId}  payload=${registered_model_update}
```
Comment (tarilabs) [inline test/robot/UserStory.robot:48]: as expected thank you 👍

**Review by tarilabs — APPROVED** [review]
thank you so much @isinyaaa 👍 🚀 

/lgtm
/approve

File: Makefile (hunk 1 above)
Comment (isinyaaa) [reply Makefile:133]: sure can look into it. I agree the Makefile is a rather "forgotten" place to specify this, as neither us nor dependabot will keep it up to date.

File: internal/converter/openapi_reconciler.go (hunk 4 above)
Comment (isinyaaa) [reply internal/converter/openapi_reconciler.go:14]: As soon as we have [copy constructors](https://github.com/jmattheis/goverter/issues/147) I intend to simplify this piece of code again, so we can think of something together.

**Pull request merged**